src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferencesSearchEngine_MapCreation.cs

```diff
@@ -23,44 +26,48 @@ private async Task<ProjectToDocumentMap> CreateProjectToDocumentMapAsync(Project
         {
             using (Logger.LogBlock(FunctionId.FindReference_CreateDocumentMapAsync, _cancellationToken))
             {
-                var finalMap = new ProjectToDocumentMap();
+                using var _ = ArrayBuilder<Task<(ImmutableArray<Document>, ISymbol, IReferenceFinder)>>.GetInstance(out var tasks);
 
-                foreach (var kvp in projectMap)
+                foreach (var (project, projectQueue) in projectMap)
                 {
-                    var project = kvp.Key;
-                    var projectQueue = kvp.Value;
-
-                    var documentMap = new DocumentMap();
-
-                    foreach (var symbolAndFinder in projectQueue)
+                    foreach (var (symbol, finder) in projectQueue)
                     {
-                        _cancellationToken.ThrowIfCancellationRequested();
+                        tasks.Add(Task.Run(() =>
+                            DetermineDocumentsToSearchAsync(project, symbol, finder), _cancellationToken));
+                    }
+                }
 
-                        var (symbol, finder) = symbolAndFinder;
+                var finalMap = new ProjectToDocumentMap();
 
-                        var documents = await finder.DetermineDocumentsToSearchAsync(
-                            symbol, project, _documents, _options, _cancellationToken).ConfigureAwait(false);
-                        foreach (var document in documents.Distinct().WhereNotNull())
+                var results = await Task.WhenAll(tasks).ConfigureAwait(false);
+                foreach (var (documents, symbol, finder) in results)
+                {
+                    foreach (var document in documents)
+                    {
+                        if (!finalMap.TryGetValue(document.Project, out var documentMap))
```

Comment:
I think this could be simplified with [GetOrAdd](http://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/IDictionaryExtensions.cs,18)
```cs
finalMap.GetOrAdd(document.Project, new DocumentMap()).Add(document, (symbol, finder));
```

Reply:
that would allocate a new DocumentMap each time.  I'd prefer to not do that :)

Reply:
Done :)